src/TimeTrace/ExecutionHierarchy.cpp

```diff
@@ -0,0 +1,342 @@
+    }
+    else
+    {
+        // keep full hierarchy when root TemplateInstantiation passes filter, even if children wouldn't pass
+        auto rootDurationMs = std::chrono::duration_cast<std::chrono::milliseconds>(templateInstantiation.Duration());
+        if (rootDurationMs < filter_.IgnoreTemplateInstantiationUnderMs) {
+            IgnoreEntry(templateInstantiation.EventInstanceId(), parent.EventInstanceId());
+        }
+        else
+        {
+            // get us subscribed for name resolution (may already have some other activities following)
+            auto result = unresolvedTemplateInstantiationsPerSymbol_.try_emplace(templateInstantiation.SpecializationSymbolKey(),
+                                                                                 TUnresolvedTemplateInstantiationNames());
+            result.first->second.push_back(templateInstantiation.EventInstanceId());
+        }
+    }
+}
+
+void ExecutionHierarchy::OnFinishNestedTemplateInstantiation(const TemplateInstantiationGroup& templateInstantiationGroup,
+                                                             const TemplateInstantiation& templateInstantiation)
+{
+    assert(templateInstantiationGroup.Size() > 0);
+
+    if (!filter_.AnalyzeTemplates) {
+        IgnoreEntry(templateInstantiation.EventInstanceId(), templateInstantiationGroup.Back().EventInstanceId());
+    }
+    else
+    {
+        // get us subscribed for name resolution (may already have some other activities following)
+        auto result = unresolvedTemplateInstantiationsPerSymbol_.try_emplace(templateInstantiation.SpecializationSymbolKey(),
+                                                                                TUnresolvedTemplateInstantiationNames());
+        result.first->second.push_back(templateInstantiation.EventInstanceId());
+    }
+}
+
+void ExecutionHierarchy::OnSymbolName(const SymbolName& symbolName)
+{
+    // SymbolName events get executed after all TemplateInstantiation in the same FrontEndPass take place
+    // and we're sure to have exclusive keys for these symbols (they may have matching names between
+    // FrontEndPass activities, but their key is unique for the trace)
+    assert(symbolNames_.find(symbolName.Key()) == symbolNames_.end());
+    std::string& name = symbolNames_[symbolName.Key()];
+    name = symbolName.Name();
+
+    // now we've resolved the name, let subscribed activities know
+    auto itSubscribedForSymbol = unresolvedTemplateInstantiationsPerSymbol_.find(symbolName.Key());
+    if (itSubscribedForSymbol != unresolvedTemplateInstantiationsPerSymbol_.end())
+    {
+        for (unsigned long long id : itSubscribedForSymbol->second)
+        {
+            auto itEntry = entries_.find(id);
+            
+            // may've been filtered out (didn't clean up the subscription as we're cleaning them all in a bit)
+            if (itEntry != entries_.end())
```

Comment:
Would this still be required if changing the `OnTemplateInstantiation` function to the version suggested above?

Reply:
There would be no problem in deleting this code if we found a way to prevent unnecessary subscriptions (or if you prefer we perform the subscription clean-up when a hierarchy is filtered out).

Reply:
Let's leave this like this for now as per comment above.